File: client-java/controller/src/main/java/org/evomaster/client/java/controller/mongo/selectors/SingleConditionQuerySelector.java
package org.evomaster.client.java.controller.mongo.selectors;

import org.evomaster.client.java.controller.mongo.operations.*;

import java.util.Map;
import java.util.Objects;
import java.util.Set;

import static org.evomaster.client.java.controller.mongo.utils.BsonHelper.*;

/**
 * Selectors for operations whose value consist of a single condition
 */
abstract class SingleConditionQuerySelector extends QuerySelector {
    @Override
    public QueryOperation getOperation(Object bsonDocument) {
        Objects.requireNonNull(bsonDocument);

        if (!(bsonDocument instanceof Map)) {
            return null;
        }

        String fieldName = extractFieldName(bsonDocument);
        if (fieldName == null) {
            return null;
        }

        if (!isUniqueEntry((Map<?, ?>) bsonDocument)) {
            return null;
        }

        Object innerDoc = getValue(bsonDocument, fieldName);
        if (!isBsonDocument(innerDoc) || !hasTheExpectedOperator(bsonDocument)) {
            return null;
        }

        Set<String> innerKeys = documentKeys(innerDoc);
        if (innerKeys == null || innerKeys.size() != 1) {
            return null;
        }

        if (innerKeys.contains(operator())) {
            Object value = getValue(innerDoc, operator());
            return parseValue(fieldName, value);
        } else {
            return null;
        }

    }

    protected String extractOperator(Object query) {
        String fieldName = extractFieldName(query);
        if (fieldName == null) return null;
        Set<String> keys = documentKeys(getValue(query, fieldName));
        return keys == null ? null : keys.stream().findFirst().orElse(null);
    }

    protected abstract QueryOperation parseValue(String fieldName, Object value);

    private String extractFieldName(Object query) {
        Set<String> keys = documentKeys(query);
        return keys == null ? null : keys.stream().findFirst().orElse(null);
    }
}

File: client-java/controller/src/main/java/org/evomaster/client/java/controller/mongo/selectors/SingleConditionQuerySelector.kt
package org.evomaster.client.java.controller.mongo.selectors

import org.bson.Document
import org.evomaster.client.java.controller.mongo.operations.*

/**
 * Selectors for operations whose value consist of a single condition
 */
abstract class SingleConditionQuerySelector : QuerySelector() {
    override fun getOperation(query: Document): QueryOperation? {
        val fieldName = extractFieldName(query)
        if (!isUniqueEntry(query) || query[fieldName] !is Document || !hasTheExpectedOperator(query)) return null
        val value = (query[fieldName] as Document)[operator()]
        return parseValue(fieldName, value)
    }

    override fun extractOperator(query: Document): String {
        val fieldName = extractFieldName(query)
        return  (query[fieldName] as Document).keys.first()
    }

    protected abstract fun parseValue(fieldName: String, value: Any?): QueryOperation?

    private fun extractFieldName(query: Document): String = query.keys.first()
}